### PR TITLE
optimization of SiStip RecHits Validation

### PR DESCRIPTION
**Pull request by ngrenz**
Optimization by deleting of unused variables, set of variables only one time and restructure of if-statements.

The outputs are supposed to be unchanged.
Any comments and suggestions are welcome.

### CODE REVIEW

**Comment by ngrenz** [comment]
@boudoul FYI

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 1]
```diff
@@ -492,26 +478,29 @@ void SiStripRecHitsValid::rechitanalysis_matched(SiStripMatchedRecHit2D const re
...
+
   if(!matched.empty()){
 
     const StripGeomDetUnit* partnerstripdet =(StripGeomDetUnit*) gluedDet->stereoDet();
     std::pair<LocalPoint,LocalVector> hitPair;
-    
 
-    for(vector<PSimHit>::const_iterator m=matched.begin(); m<matched.end(); m++){
-      SiStripDetId hitDetId(m->detUnitId());
+    for(auto m : matched){
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: wrong.
use  auto const &
auto makes a copy

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 2]
```diff
@@ -492,26 +478,29 @@ void SiStripRecHitsValid::rechitanalysis_matched(SiStripMatchedRecHit2D const re
...
+      SiStripDetId hitDetId(m.detUnitId());
       if (hitDetId.stereo()) {  // project from the stereo sensor
       //project simhit;
-	hitPair= projectHit((*m),partnerstripdet,gluedDet->surface());
-	distx = fabs(rechitpro.x - hitPair.first.x());
-	disty = fabs(rechitpro.y - hitPair.first.y());
+	hitPair= projectHit(m,partnerstripdet,gluedDet->surface());
+	distx = rechitpro.x - hitPair.first.x();
+	disty = rechitpro.y - hitPair.first.y();
 	dist = sqrt(distx*distx+disty*disty);
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: as you are optimizing
please use dist2 = distx*distx+disty*disty

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 3]
```diff
@@ -420,14 +408,17 @@ void SiStripRecHitsValid::rechitanalysis(SiStripRecHit2D const rechit,const Stri
...
   PSimHit closest;
+ 
+  rechitpro.resx = -999999.; rechitpro.resy = -999999.;rechitpro.pullMF = -999999.; 
+  rechitpro.chi2 = -999999.; rechitpro.bunch = -999999.; rechitpro.event = -999999.;
   
   if(!matched.empty()){
 
-    for(vector<PSimHit>::const_iterator m=matched.begin(); m<matched.end(); m++){
-      dist = fabs(rechitpro.x - (*m).localPosition().x());
+    for(auto m : matched){
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: see below: why this code is written twice?

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 4]
```diff
@@ -420,14 +408,17 @@ void SiStripRecHitsValid::rechitanalysis(SiStripRecHit2D const rechit,const Stri
   double mindist = 999999;
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: use floats

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 5]
```diff
@@ -594,26 +582,30 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
     std::string label = hidmanager.getSubdetid(detid,tTopo,true);
     // std::cout << "label " << label << endl;
       
-    std::map<std::string, LayerMEs>::iterator iLayerME  = LayerMEsMap.find(label);
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: do not understand why getSubdetid returns a string and not a enumerator
if possible migrate all this code to use enumerators

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 6]
```diff
@@ -594,26 +582,30 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
...
+    if(LayerMEsMap.find(label)==LayerMEsMap.end()) {
 	
       // get detids for the layer
       // Keep in mind that when we are on the TID or TEC we deal with rings not wheel 
       int32_t lnumber = det_layer_pair.second;
+      std::string lname = det_layer_pair.first; 
       std::vector<uint32_t> layerDetIds;        
-      if (det_layer_pair.first == "TIB") {
-	substructure.getTIBDetectors(activeDets,layerDetIds,lnumber,0,0,0);
-      } else if (det_layer_pair.first == "TOB") {
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: never compare string with a const char *  the compiler will construct a string each time...
please either use enumerator of compare with constexpr strings

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 7]
```diff
@@ -594,26 +582,30 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
...
+      if (lname == "TEC") {
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: never compare a string with a const char * : the compiler usually construct a temporary string
use either an enumerator or a constexpr string

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 8]
```diff
@@ -466,19 +457,14 @@ void SiStripRecHitsValid::rechitanalysis(SiStripRecHit2D const rechit,const Stri
...
   
-  rechitpro.x = -999999.; rechitpro.y = -999999.; rechitpro.z = -999999.; rechitpro.resolxx = -999999.; rechitpro.resolxy = -999999.; 
-  rechitpro.resolyy = -999999.; rechitpro.resx = -999999.; rechitpro.resy = -999999.;rechitpro.pullMF = -999999.; 
-  rechitpro.clusiz = -999999.; rechitpro.cluchg = -999999.; rechitpro.chi2 = -999999.; rechitpro.NsimHit = -999999.;
-  rechitpro.bunch = -999999.; rechitpro.event = -999999.;
-
   LocalPoint position=rechit.localPosition();
   LocalError error=rechit.localPositionError();
 
   rechitpro.x = position.x();
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: what about encapulating this in a constructor of RecHitProperties?

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 9]
```diff
@@ -577,10 +566,9 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
   // std::cout << "curfold " << curfold << std::endl;
 
   createTotalMEs(ibooker);
-
   // loop over detectors and book MEs
   edm::LogInfo("SiStripTkRecHits|SiStripRecHitsValid")<<"nr. of activeDets:  "<<activeDets.size();
-  for(std::vector<uint32_t>::iterator detid_iterator = activeDets.begin(); detid_iterator!=activeDets.end(); detid_iterator++){
+  for(std::vector<uint32_t>::iterator detid_iterator=activeDets.begin(), detid_end=activeDets.end(); detid_iterator!=detid_end; ++detid_iterator){
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: be consistent, use auto as above

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 10]
```diff
@@ -624,40 +616,47 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
...
-      } else if ( (det_layer_pair.first == "TID") && (stereolnumber < 0) && (tTopo->tidIsStereo(detid) == 1) ) {
-	substructure.getTIDDetectors(activeDets,stereoandmatchedDetIds,1,0,abs(stereolnumber),1);
-	isStereo = true;
-      } else if ( (det_layer_pair.first == "TEC") && (stereolnumber > 0) && (tTopo->tecIsStereo(detid) == 1) ) {
-	substructure.getTECDetectors(activeDets,stereoandmatchedDetIds,2,0,0,0,abs(stereolnumber),1);
-	isStereo = true;
-      } else if ( (det_layer_pair.first == "TEC") && (stereolnumber < 0) && (tTopo->tecIsStereo(detid) == 1) ) {
-	substructure.getTECDetectors(activeDets,stereoandmatchedDetIds,1,0,0,0,abs(stereolnumber),1);
-	isStereo = true;
+      if ( tTopo->tibIsStereo(detid) == 1 ) {
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: if tibIsStereo(detid) tidIsStereo(detid) tecIsStereo(detid) are all identical, why not just isStereo()????
if not the case than this code is wrong

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 11]
```diff
@@ -624,40 +616,47 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
...
+        int32_t stereolnumber = det_layer_pair.second;
+        std::string stereolname = det_layer_pair.first;
+        if ( stereolname == "TEC" ) {
+          if ( stereolnumber > 0 ) {
+	    substructure.getTECDetectors(activeDets,stereoandmatchedDetIds,2,0,0,0,abs(stereolnumber),1);
+	    isStereo = true;
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: outside the "case"

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc [hunk 12]
```diff
@@ -624,40 +616,47 @@ void SiStripRecHitsValid::createMEs(DQMStore::IBooker & ibooker,const edm::Event
...
+          } else if ( stereolnumber < 0 ) {
+	    substructure.getTECDetectors(activeDets,stereoandmatchedDetIds,1,0,0,0,abs(stereolnumber),1);
```
Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: better  to use 
(stereolnumber < 0) ? 1:2,
so is much clearer what changed and what not

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc (hunk 10 above)
Comment (ngrenz) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: first thanks for the help

if I am not wrong, so the test here is in relating to the given situation of tec tid tob and tib.
I was not sure if a general test with isStereo() in combination with the test for the layer is realy sufficient, therefore I used TibIsStereo(detid), tcIsStereo(detid) ....

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc (hunk 11 above)
Comment (ngrenz) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: If you mean to put isStereo = true outside of the case, then you get a problem with stereolnumber = 0

Comment (VinInn) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: skip on it globally...

File: Validation/TrackerRecHits/src/SiStripRecHitsValid.cc (hunk 3 above)
Comment (ngrenz) [inline Validation/TrackerRecHits/src/SiStripRecHitsValid.cc]: There are small differences in calculation of the distance and error for the rechitsanalysis and rechitsanalysis_matched

**Pull request closed without merging**